addons/hw_l10n_eg_eta/controllers/main.py

```diff
@@ -0,0 +1,195 @@
+import hashlib
+from datetime import datetime
+
+import pkcs11
+import platform
+import base64
+import jinja2
+import json
+import pytz
+import requests
+import os
+import sys
+from pkcs11 import Attribute, ObjectClass
+from asn1crypto import cms, x509, algos, core, tsp
+from pkcs11.exceptions import NoSuchToken, PinIncorrect, MultipleTokensReturned
+
+from odoo import http
+
+
+if hasattr(sys, 'frozen'):
+    # When running on compiled windows binary, we don't have access to package loader.
+    path = os.path.realpath(os.path.join(os.path.dirname(__file__), '..', 'views'))
+    loader = jinja2.FileSystemLoader(path)
+else:
+    loader = jinja2.PackageLoader('odoo.addons.hw_l10n_eg_eta', "views")
+
+jinja_env = jinja2.Environment(loader=loader, autoescape=True)
+jinja_env.filters["json"] = json.dumps
+
+login_template = jinja_env.get_template('login.html')
+success_template = jinja_env.get_template('success.html')
+error_template = jinja_env.get_template('error.html')
+
+
+class EtaUsbController(http.Controller):
+
+    @http.route('/hw_drivers/sign/login', type='http', auth='none', cors='*', csrf=False, save_session=False)
+    def eta_sign_login(self, url, otp, invoice_ids, user):
+        return login_template.render({
+            'url': url,
+            'otp': otp,
+            'invoice_ids': invoice_ids,
+            'user': user
+        })
+
+    @http.route('/hw_drivers/sign', type='http', auth='none', cors='*', csrf=False, save_session=False)
+    def eta_sign(self, url, otp, invoice_ids, pin, user):
+        try:
+            payload = {'token': otp, 'user': user, 'invoice_ids': invoice_ids}
+            invoices = requests.get(url + '/l10n_eg_invoice/get_invoice', params=payload)
+
+            invoices = invoices.json()
+            if invoices.get('errors'):
+                return self.get_error_template(invoices.get('errors'))
+
+            invoice_dict = dict()
+
+            lib = pkcs11.lib(self.get_crypto_lib())
+            try:
+                # some manufacturer have their name in the label, some as manufacturer_id some don't have any, so we ask
+                # the user to only have the thumb drive for signing
+                token = lib.get_token()
+            except NoSuchToken:
+                return self.get_error_template('No drive found', 'Make sure the thumb drive is correctly inserted')
+            except MultipleTokensReturned:
+                return self.get_error_template('Multiple drive detected', 'Only one secure thumb drive can be inserted at the same time')
+            except PinIncorrect:
+                return self.get_error_template('Incorrect pin')
+
+            with token.open(user_pin=pin) as session:
+                cert = next(session.get_objects({Attribute.CLASS: ObjectClass.CERTIFICATE}))
+                priv_key = next(session.get_objects({Attribute.CLASS: ObjectClass.PRIVATE_KEY, Attribute.ID: cert[Attribute.ID]}))
```

Comment:
The private key is marked as unexportable so if we try to take a dump of it or read the exponent part (which is the only part hidden), it will throw an error. I guess that when we call the sign function, it will actually sign on the usb as to prevent us to ever know it. ODO can probably provide a better "why" explanation